architecture/2025-06-upgrades-gitops.md

```diff
@@ -0,0 +1,127 @@
+# Radius Upgrades with GitOps
+
+## Problem
+Radius control plane upgrades require manual CLI commands (`rad upgrade kubernetes`, `rad rollback kubernetes`). These commands wrap `helm upgrade` and `helm rollback` with additional logic, such as preflight checks (and in the future, locks). This additional logic should ideally run however the user chooses to upgrade Radius, including `rad` CLI, `helm` CLI, or GitOps-based workflows such as `Flux` and `ArgoCD`. 
+
+## Current Upgrade System
+- `rad upgrade kubernetes` - CLI upgrade command with preflight checks
+- `rad rollback kubernetes` - CLI rollback to previous/specific revision
+- `pkg/upgrade/preflight/` - Validation framework (connectivity, versions, resources, etc.)
+
+## Options
+
+### Option 1: Helm Pre-Upgrade Job
+
+**How**: Helm hook job runs preflight checks before upgrade starts.
+
+**Context**: [Helm hooks](https://helm.sh/docs/topics/charts_hooks/) run at specific lifecycle points. Pre-upgrade hooks execute before chart resources are updated, and job failure blocks the entire upgrade.
+
+ **Real-world example**: [Kong's Helm chart](https://github.com/Kong/charts/blob/main/charts/kong/templates/migrations-pre-upgrade.yaml) uses this pattern for database migrations before upgrades.
+
+**Implementation**: 
+- Create Job template with `helm.sh/hook: pre-install,pre-upgrade`
+- Package existing `pkg/upgrade/preflight` code from `pkg/upgrade/preflight/registry.go`
```

Comment:
This will need to have a main.go and Dockerfile. I feel like it may be over-engineering at this point of `rad upgrade kubernetes` and `rad rollback kubernetes`.